crates/sui-framework/tests/linked_table_tests.move

```diff
@@ -0,0 +1,372 @@
+    let table = linked_table::new<u64, u64>(ts::ctx(&mut scenario));
+    borrow(&table, 0);
+    abort 42
+}
+
+#[test]
+#[expected_failure(abort_code = sui::dynamic_field::EFieldDoesNotExist)]
+fun borrow_mut_missing() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new<u64, u64>(ts::ctx(&mut scenario));
+    borrow_mut(&mut table, 0);
+    abort 42
+}
+
+#[test]
+#[expected_failure(abort_code = sui::dynamic_field::EFieldDoesNotExist)]
+fun remove_missing() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new<u64, u64>(ts::ctx(&mut scenario));
+    remove(&mut table, 0);
+    abort 42
+}
+
+#[test]
+#[expected_failure(abort_code = linked_table::ETableIsEmpty)]
+fun pop_front_empty() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new<u64, u64>(ts::ctx(&mut scenario));
+    pop_front(&mut table);
+    abort 42
+}
+
+#[test]
+#[expected_failure(abort_code = linked_table::ETableIsEmpty)]
+fun pop_back_empty() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new<u64, u64>(ts::ctx(&mut scenario));
+    pop_back(&mut table);
+    abort 42
+}
+
+#[test]
+#[expected_failure(abort_code = linked_table::ETableNotEmpty)]
+fun destroy_non_empty() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new(ts::ctx(&mut scenario));
+    push_back(&mut table, 0, 0);
+    destroy_empty(table);
+    ts::end(scenario);
+}
+
+#[test]
+fun sanity_check_contains() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new(ts::ctx(&mut scenario));
+    assert!(!contains(&mut table, 0), 0);
+    push_back(&mut table, 0, 0);
+    assert!(contains<u64, u64>(&mut table, 0), 0);
+    assert!(!contains<u64, u64>(&mut table, 1), 0);
+    ts::end(scenario);
+    drop(table);
+}
+
+#[test]
+fun sanity_check_drop() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new(ts::ctx(&mut scenario));
+    push_back(&mut table, 0, 0);
+    assert!(length(&table) == 1, 0);
+    ts::end(scenario);
+    drop(table);
+}
+
+#[test]
+fun sanity_check_size() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let table = linked_table::new(ts::ctx(&mut scenario));
+    assert!(is_empty(&table), 0);
+    assert!(length(&table) == 0, 0);
+    push_back(&mut table, 0, 0);
+    assert!(!is_empty(&table), 0);
+    assert!(length(&table) == 1, 0);
+    push_back(&mut table, 1, 0);
+    assert!(!is_empty(&table), 0);
+    assert!(length(&table) == 2, 0);
+    ts::end(scenario);
+    drop(table);
+}
+
+#[test]
+fun test_all_orderings() {
+    let sender = @0x0;
+    let scenario = ts::begin(sender);
+    let ctx = ts::ctx(&mut scenario);
+    let keys = vector[b"a", b"b", b"c"];
+    let values = vector[3, 2, 1];
+    let all_bools = vector[
```

Comment:
This tests all possible orderings of push_fronts/push_backs followed by all possible orderings of pop_front/pop_back and checks that the ordering is correct. 

I wish I could figure out a way to also take from the middle but it felt difficult to then also check the ordering.